xbmc/platform/posix/storage/discs/DiscDriveHandlerPosix.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ *  Copyright (C) 2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DiscDriveHandlerPosix.h"
+
+#include "storage/cdioSupport.h"
+#include "utils/log.h"
+
+#include <memory>
+
+namespace
+{
+constexpr int MaxOpenRetries = 3;
+}
+using namespace MEDIA_DETECT;
+
+
+std::shared_ptr<IDiscDriveHandler> IDiscDriveHandler::CreateInstance()
+{
+  return std::make_shared<DiscDriveHandlerPosix>();
+}
+
+DriveState DiscDriveHandlerPosix::GetDriveState(const std::string& devicePath)
+{
+  const std::shared_ptr<CLibcdio> libCdio = CLibcdio::GetInstance();
+  if (!libCdio)
+  {
+    CLog::LogF(LOGERROR, "Failed to obtain libcdio handler");
+    return DriveState::DRIVE_NOT_READY;
+  }
+
+  DriveState driveStatus = DriveState::DRIVE_NOT_READY;
+  CdIo_t* cdio = libCdio->cdio_open(devicePath.c_str(), DRIVER_UNKNOWN);
+  if (!cdio)
+  {
+    CLog::LogF(LOGERROR, "Failed to open device {} libcdio handler", devicePath);
+    return DriveState::DRIVE_NOT_READY;
+  }
+
+  CdioTrayStatus status = static_cast<CdioTrayStatus>(libCdio->mmc_get_tray_status(cdio));
+
+  switch (status)
+  {
+    case CdioTrayStatus::CLOSED:
+      driveStatus = DriveState::DRIVE_CLOSED_MEDIA_UNDEFINED;
+      break;
+
+    case CdioTrayStatus::OPEN:
+      driveStatus = DriveState::DRIVE_OPEN;
+      break;
+
+    default:
+      CLog::LogF(LOGWARNING, "Unknown/unhandled drive state interpreted as DRIVE_NOT_READY");
+      break;
+  }
+  libCdio->cdio_destroy(cdio);
+
+  return driveStatus;
+}
+
+TrayState DiscDriveHandlerPosix::GetTrayState(const std::string& devicePath)
+{
+  const std::shared_ptr<CLibcdio> libCdio = CLibcdio::GetInstance();
+  if (!libCdio)
+  {
+    CLog::LogF(LOGERROR, "Failed to obtain libcdio handler");
+    return TrayState::UNDEFINED;
+  }
+
+  TrayState trayStatus = TrayState::UNDEFINED;
```

Comment:
you could move decl of this var to the beginning of this method, init it with UNDEFINED and return this var everywhere in this method intsead of sometimes returning UNDEFINED derictly sometimes the var's value